dnf-json

```diff
@@ -1,95 +1,25 @@
 #!/usr/bin/python3
```

Comment:
By moving the state out of `dnf-json` and into the go wrapper, do we lose the ability to reuse state between calls, or was that never going to be possible anyway? (i.e., will we always need to re-read state from disk?)

Reply:
The current implementation of `dnf-json` needs to fork on every call because of memory leaks, so there's no state reuse right now either.  That's partially the reason for this redesign.  The idea is to make `dnf-json` a much simpler "dumb" binary and manage everything from the Go side.  It's a bit wasteful because we're starting up the Python interpreter with every call, but it seems there's no getting around that anyway.

Cache invalidation and cleanup will happen on the Go side as well and that's planned for this PR.  For now, it can just replicate what `dnf-json` does but we could also address https://github.com/osbuild/osbuild-composer/issues/2446.

Reply:
I was thinking the other day, what if the DNF api leaks because we open install transactions that we never close ? Maybe there's a way to cancel an install transaction without having to really install the packages.

Reply:
Right, what I meant was the on-disk state. I think an interesting question is if it would be possible to solve the problem of somehow preserving in-memory state, but if that's not going to be possible then doing it all in go rather than python sounds good.

The reason I bring that up is that depsolving is currently taking surprisingly long, even when the caches should be hot. Either we are doing something wrong and not actually reusing the caches properly, or it really just takes a very long time to load them from disk (in which case trying to keep the thing in memory may make sense). Any of this does not have to be solved by this PR, just something to keep in mind for deciding the direction to take.

Related to this, what we could do would be to support passing several depsolve requests in one transaction (not the chained ones you already plan for, but really distinct one: for the build root, for the payload, etc), and in that case reuse state for the full transaction, so at least we only load from disk once per call to `compose`.

Reply:
> Related to this, what we could do would be to support passing several depsolve requests in one transaction (not the chained ones you already plan for, but really distinct one: for the build root, for the payload, etc), and in that case reuse state for the full transaction, so at least we only load from disk once per call to `compose`.

This can definitely be supported.  With the changes in this PR, we could also theoretically do all the depsolves for a given build asynchronously (one go routine per package set, each spawning a separate `dnf-json`), which isn't possible with the current service-based one.  They would all hit the same cache, which `dnf` locks, but it'd probably result in a speedup.